Fix invalid SQL generated for bulk upsert with empty array

Qu4tro opened the pull request:
Simple fix and respective test that resolves #1265.

I've added a simple check to check what columns are selected. If none are (the reason behind the syntax error), then it creates a `DO NOTHING` as suggested in the issue.

steve-chavez commented on test/Feature/UpsertSpec.hs:
```diff
@@ -44,6 +44,12 @@ spec =
+        it "succeeds when the payload has no elements" $
+          request methodPost "/articles" [("Prefer", "return=representation"), ("Prefer", "resolution=merge-duplicates")]
+            [json|[]|] `shouldRespondWith`
+            [json|[]|] { matchStatus = 201 , matchHeaders = [matchContentTypeJson] }
```
HTTP 201 Created is a bit misleading, an HTTP 200 OK would be better. You could conditionally change the status by comparing if `queryTotal = 0` I believe.

Qu4tro replied:
I agree with you that HTTP 201 is misleading.

I've tried to not to de-rail, from current behavior and I had found this test:
```
-- test/Feature/InsertSpec.hs (lines 131-137)
it "returns empty array when no items inserted, and return=rep" $ do
  p <- request methodPost "/no_pk"
               [("Prefer", "return=representation")]
               [json| [] |]
  liftIO $ do
    simpleBody p `shouldBe` [json| [] |]
    simpleStatus p `shouldBe` created201
```

Which I would argue is similar to the one created in this PR. 
Let me know, which option do you think it's more fitting: Change both behaviors to 200, change just this PR to 200 or keep it as is for consistency.

steve-chavez replied:
Both behaviors to 200 seems correct. My reasoning is that with [201](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/201) we can include a `Location` header for identifying the resource, and with `[]` as the response we can't do that.

steve-chavez commented on src/PostgREST/App.hs:
```diff
@@ -197,9 +197,10 @@ app dbStructure proc conf apiRequest =
+                    s_200 = if iPreferRepresentation apiRequest == Full
                           then status200
                           else status204
+                    s = if queryTotal == 0 then status404 else s_200
```
Is this related to https://github.com/PostgREST/postgrest/issues/1257?

Qu4tro replied:
Oh. That was a mistake, wrong branch. Sorry about that.

Yes. I was trying to fix it, but my current approach, breaks a few other tests.

I can create a PR and we can discuss it there.

steve-chavez replied:
Thank you. That makes reviewing easier.

steve-chavez commented:
You could avoid the test timeouts by rebasing to master.

Qu4tro commented:
Changed return code to 200 and both tests to expect 200 as discussed.

A few others tests started to fail as a result, which I changed them as well.
These fall under two fields:

`return=minimal` - When using return=minimal, the query that it is built is hardcoded with a zero. I'm guessing that's for performance reasons and I don't think that's bad, since it's explicit. 
`resolution=ignore-duplicates` when no row is actually inserted. 

I think both cases are fitted for a 200 as well. In this second case, no resource is created, similarly to the case in the original discussion.

In the `return=minimal` the request is already being handicapped and we don't actually know if a resource was created (i.e. ignore duplicates above), so I think a 200 OK (we ran the query and everything went fine) is appropriate.

steve-chavez commented:
@Qu4tro Just realized I bumped into the `return=minimal` problem in a previous PR https://github.com/PostgREST/postgrest/pull/1048#discussion_r169307092, opened an issue for that here https://github.com/PostgREST/postgrest/issues/1070.

I think `return=minimal` responding with 200 is not accurate when a resource is created, it should be `201`. We could make it work by having the same query for this [two](https://github.com/PostgREST/postgrest/blob/master/src/PostgREST/QueryBuilder.hs#L111-L117) and returning [xmax](https://www.postgresql.org/docs/11/ddl-system-columns.html) in this [line](https://github.com/PostgREST/postgrest/blob/master/src/PostgREST/DbRequestBuilder.hs#L342).

This is getting away from the original https://github.com/PostgREST/postgrest/issues/1265 though. So how about if you open another PR for #1070?
You could remove your last commit and then I'll merge this PR.

steve-chavez commented:
Also, don't forget to add an entry to the [CHANGELOG](https://github.com/PostgREST/postgrest/blob/master/CHANGELOG.md).

Qu4tro commented:
Cool, all done.
